pkgs/tools/virtualization/nixos-nspawn/nixos-nspawn.py

```diff
@@ -0,0 +1,512 @@
+            for line in f:
+                if line.startswith('#'):
+                    continue
+                match = ptrn.match(line)
+                if not match:
+                    continue
+                data[match.group('prop')] = match.group('value').strip()
+
+        if data == {}:
+            return '<unknown>'
+
+        self.version = f"{data['NAME']} {data['VERSION']}"
+
+        return self.version
+
+    def dump(self):
+        return {
+            'kind': self.type(),
+            'state': self.__machinectl_property('State', 'powered off'),
+            'name': self.name,
+            'file': self.path,
+            'os': self.version
+        }
+
+    def type(self):
+        return 'imperative' if self.is_imperative() else 'declarative'
+
+    def list_generations(self):
+        profile_dir = self.__profile_dir()
+        run([
+            "nix-env",
+            "-p",
+            f"{profile_dir}/system",
+            "--list-generations"
+        ], log_raw=True)
+
+    def rollback(self, strategy=None):
+        data = self.__cfg_repr()
+
+        profile_dir = self.__profile_dir()
+        run([
+            "nix-env",
+            "-p",
+            f"{profile_dir}/system",
+            "--rollback"
+        ], log_raw=True)
+
+        self.__activate(
+            data=data,
+            path=self.__nix_path(),
+            strategy=strategy
+        )
+
+    def remove(self):
+        logging.info(f"Removing {self.name}...")
+        run(["machinectl", "poweroff", self.name], must_succeed=False)
+
+        logging.info(f"Removing nspawn unit {self.path}")
+        try:
+            remove(self.path)
+            remove(f'/etc/systemd/network/20-ve-{self.name}.network')
+        except FileNotFoundError:
+            logging.warn(f"{self.path} or network unit doesn't exist! Continuing with cleanup anyways")
+
+        rmtree(f"{PROFILE_BASE}/{self.name}", ignore_errors=True)
+
+        logging.info(f"Removing state-directory in /var/lib/machines/{self.name}")
+        rmtree(f"/var/lib/machines/{self.name}", ignore_errors=True)
+
+    def build_nixos_config(self, nix_expr, update=False, show_trace=False):
+        if not update:
+            try:
+                profile_dir = setup_nixprofile(self.name)
+            except FileExistsError:
+                raise Exception(
+                    f"Profile for {self.name} already exists! Perhaps some dirty state? Try removing with `remove'."
+                )
+        else:
+            profile_dir = self.__profile_dir()
+
+        eval_code = getenv('NIXOS_NSPAWN_EVAL', '@eval@')
+        trace_arg = None if not show_trace else "--show-trace"
+        args = [
+            "nix-env", "-p", f"{profile_dir}/system", "--arg", "config", nix_expr,
+            "-f", eval_code, "--set", "--arg", "nixpkgs", "<nixpkgs>"
+        ]
+        if trace_arg is not None:
+            args.append(trace_arg)
+        run(args)
```

Comment:
for me this doesnt seem to print output by default which meant I thought it hang.